eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/CDDL.hs

```diff
@@ -0,0 +1,404 @@
+  module Test.Cardano.Ledger.Babbage.CDDL,
+) where
+
+import Codec.CBOR.Cuddle.Huddle
+import Data.Function (($))
+import Data.Word (Word64)
+import GHC.Num (Integer)
+import Test.Cardano.Ledger.Alonzo.CDDL hiding (
+  auxiliary_data,
+  block,
+  cost_models,
+  header,
+  header_body,
+  language,
+  major_protocol_version,
+  multiasset,
+  next_major_protocol_version,
+  plutus_data,
+  plutus_script,
+  proposed_protocol_parameter_updates,
+  protocol_param_update,
+  protocol_version,
+  script_data_hash,
+  script_n_of_k,
+  transaction,
+  transaction_body,
+  transaction_output,
+  transaction_witness_set,
+  update,
+ )
+import Text.Heredoc
+
+babbageCDDL :: Huddle
+babbageCDDL = collectFrom [block, transaction, kes_signature, language, signkeyKES]
+
+block :: Rule
+block =
+  comment
+    [str|Valid blocks must also satisfy the following two constraints:
+        |  1) the length of transaction_bodies and transaction_witness_sets must be
+        |     the same
+        |  2) every transaction_index must be strictly smaller than the length of
+        |     transaction_bodies
+        |NEW: 
+        |  invalid_transactions
+        |]
+    $ "block"
+      =:= arr
+        [ a header
+        , "transaction_bodies" ==> arr [0 <+ a transaction_body]
+        , "transaction_witness_sets" ==> arr [0 <+ a transaction_witness_set]
+        , "auxiliary_data_set" ==> mp [0 <+ asKey transaction_index ==> auxiliary_data]
+        , "invalid_transactions" ==> arr [0 <+ a transaction_index]
+        ]
+
+transaction :: Rule
+transaction =
+  "transaction"
+    =:= arr
+      [ a transaction_body
+      , a transaction_witness_set
+      , a VBool
+      , a (auxiliary_data / VNil)
+      ]
+
+header :: Rule
+header = "header" =:= arr [a header_body, "body_signature" ==> kes_signature]
+
+header_body :: Rule
+header_body =
+  comment
+    [str| block_body_size: merkle triple root
+        |      vrf_result: NEW, replaces nonce_vrf and leader_vrf
+        |]
+    $ "header_body"
+      =:= arr
+        [ "block_number" ==> VUInt
+        , "slot" ==> VUInt
+        , "prev_hash" ==> (hash32 / VNil)
+        , "issuer_vkey" ==> vkey
+        , "vrf_vkey" ==> vrf_vkey
+        , "vrf_result" ==> vrf_cert
+        , "block_body_size" ==> VUInt
+        , "block_body_hash" ==> hash32
+        , a operational_cert
+        , a protocol_version
+        ]
+
+-- TODO Replace with the following once
+-- https://github.com/input-output-hk/cuddle/issues/29 is addressed in cuddle.
+--
+-- next_major_protocol_version :: Rule
+-- next_major_protocol_version = "next_major_protocol_version" =:= (7 :: Integer)
+next_major_protocol_version :: Integer
+next_major_protocol_version = 9
+
+major_protocol_version :: Rule
+major_protocol_version = "major_protocol_version" =:= (1 :: Integer) ... next_major_protocol_version
+
+-- FIXME: This is wrong.
+protocol_version :: Rule
+protocol_version = "protocol_version" =:= arr [a major_protocol_version, a VUInt]
```

Comment:
There should be a tuple here inside the array, is this what your FIXME comment above refers to? 
I reckon we can't merge this before it's fixed

Is it difficult?  Isn't it just a matter of having the array over the tuple that was defined in alonzo? `grp [a major_protocol_version, a VUInt]`   Or is there more to the problem?

Reply:
It is a stale comment because I just repeated what is done in `conway`. From what I read, this is because CBOR doesn't have a direct way to represent a "group" as defined by CDDL parentheses. The parentheses in CDDL are a syntactic grouping mechanism rather than a data structure specification. It is very confusing actually, to use a group as the element of a 1-element list. The tests pass because the two definitions are equivalent for _our_ purposes, though they are actually different in theory.

Specifically, groups delimited by parentheses are also just encoded as arrays in CBOR. So `[(a,b)]` will be encoded as `array of an array containing a and b`, whereas, `[a,b]` or `(a, b)` both will be encoded as `array containing a and b`.

Please correct me if I'm wrong @lehins.

Reply:
I see. Interesting, thanks!